In [3]:
install.packages("tidyverse")

also installing the dependencies 'colorspace', 'sys', 'bit', 'ps', 'sass', 'cachem', 'farver', 'labeling', 'munsell', 'RColorBrewer', 'viridisLite', 'rappdirs', 'rematch', 'askpass', 'bit64', 'prettyunits', 'processx', 'highr', 'xfun', 'yaml', 'bslib', 'fontawesome', 'jquerylib', 'tinytex', 'backports', 'generics', 'memoise', 'blob', 'DBI', 'R6', 'tidyselect', 'withr', 'data.table', 'gtable', 'isoband', 'scales', 'gargle', 'cellranger', 'curl', 'ids', 'rematch2', 'cpp11', 'pkgconfig', 'mime', 'openssl', 'timechange', 'systemfonts', 'textshaping', 'clipr', 'vroom', 'tzdb', 'progress', 'callr', 'fs', 'knitr', 'rmarkdown', 'selectr', 'stringi', 'broom', 'conflicted', 'dbplyr', 'dplyr', 'dtplyr', 'forcats', 'ggplot2', 'googledrive', 'googlesheets4', 'haven', 'hms', 'httr', 'lubridate', 'magrittr', 'modelr', 'purrr', 'ragg', 'readr', 'readxl', 'reprex', 'rstudioapi', 'rvest', 'stringr', 'tibble', 'tidyr', 'xml2'




package 'colorspace' successfully unpacked and MD5 sums checked
package 'sys' successfully unpacked and MD5 sums checked
package 'bit' successfully unpacked and MD5 sums checked
package 'ps' successfully unpacked and MD5 sums checked
package 'sass' successfully unpacked and MD5 sums checked
package 'cachem' successfully unpacked and MD5 sums checked
package 'farver' successfully unpacked and MD5 sums checked
package 'labeling' successfully unpacked and MD5 sums checked
package 'munsell' successfully unpacked and MD5 sums checked
package 'RColorBrewer' successfully unpacked and MD5 sums checked
package 'viridisLite' successfully unpacked and MD5 sums checked
package 'rappdirs' successfully unpacked and MD5 sums checked
package 'rematch' successfully unpacked and MD5 sums checked
package 'askpass' successfully unpacked and MD5 sums checked
package 'bit64' successfully unpacked and MD5 sums checked
package 'prettyunits' successfully unpacked and MD5 sums checked
package 'processx' success

In [7]:
# Importing necessary libraries
library("tidyverse")
library("ggplot2")
library("ggrepel")
library("ggcorrplot")
library("DT")

── Attaching core tidyverse packages ──────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


ERROR: Error in library("ggrepel"): there is no package called 'ggrepel'


# Dataset : COVID-19_cases_plus_census (updated)

In [27]:
dataset_plus_census <- read.csv("https://raw.githubusercontent.com/yamphurai/Project2_ClusterAnalysis/refs/heads/main/updated_COVID-19_cases_plus_census.csv")

In [28]:
# First 10 rows
head(dataset_plus_census)

state state_fips_code county_name       confirmed_cases deaths total_pop
1 VT    50              Essex County        111             0      6203   
2 VT    50              Chittenden County  3636            78    160985   
3 DE    10              Kent County       11548           187    173145   
4 RI    44              Washington County  5521           122    126190   
5 NH    33              Belknap County     2496            79     60383   
6 RI    44              Newport County     3578             6     83204   
  male_pop female_pop white_pop black_pop ⋯ male_21_49 male_50_above
1  3135     3068        5929       64     ⋯   900       1701        
2 78928    82057      143657     4091     ⋯ 32947      27110        
3 83544    89601      108627    41729     ⋯ 31121      29975        
4 61154    65036      115206     1621     ⋯ 20438      27238        
5 29705    30678       57523      285     ⋯  9663      14520        
6 40952    42252       71549     2596     ⋯ 15262      18187        
  female_0_20 female_21_49 female_50_above unemployed_pop employed_pop commute
1   603         924         1657            149            2784          4820 
2 20668       32285        30864           3827           90054        148357 
3 23801       33109        35761           5563           78078        145421 
4 16106       20580        30387           4142           64576        114755 
5  6541        9882        15317           1589           30674         55288 
6  9040       14232        20449           2349           41471         73711 
  worked_at_home walked_to_work
1  167             81          
2 4783           6859          
3 3270           1379          
4 3110           1725          
5 1487            449          
6 2711           2749

In [30]:
# Checking for missing values
sum(is.na(dataset_plus_census))

[1] 0

In [31]:
# Basic statistics
summary(dataset_plus_census)

    state           state_fips_code county_name        confirmed_cases    
 Length:3142        Min.   : 1.00   Length:3142        Min.   :      0.0  
 Class :character   1st Qu.:18.00   Class :character   1st Qu.:    796.2  
 Mode  :character   Median :29.00   Mode  :character   Median :   1916.5  
                    Mean   :30.28                      Mean   :   7558.9  
                    3rd Qu.:45.00                      3rd Qu.:   4955.0  
                    Max.   :56.00                      Max.   :1002614.0  
     deaths          total_pop           male_pop         female_pop     
 Min.   :    0.0   Min.   :      74   Min.   :     39   Min.   :     35  
 1st Qu.:   12.0   1st Qu.:   10945   1st Qu.:   5514   1st Qu.:   5460  
 Median :   31.0   Median :   25692   Median :  12798   Median :  12885  
 Mean   :  124.8   Mean   :  102166   Mean   :  50292   Mean   :  51873  
 3rd Qu.:   77.0   3rd Qu.:   67445   3rd Qu.:  33481   3rd Qu.:  34108  
 Max.   :13936.0   Max.   :1010

## Scale the data

In [16]:
# Prepare data
dataset_cases_TX_clustering <- dataset_cases_TX %>%
  select(county_name, confirmed_cases)

In [17]:
# Scale the confirmed_cases column for K-means
dataset_cases_TX_clustering$scaled_confirmed_cases <- scale(dataset_cases_TX_clustering$confirmed_cases)

In [29]:
# internal structure of the dataset
str (dataset_plus_census)

'data.frame':	3142 obs. of  33 variables:
 $ state                : chr  "VT" "VT" "DE" "RI" ...
 $ state_fips_code      : int  50 50 10 44 33 44 50 9 50 50 ...
 $ county_name          : chr  "Essex County" "Chittenden County" "Kent County" "Washington County" ...
 $ confirmed_cases      : int  111 3636 11548 5521 2496 3578 312 6255 527 307 ...
 $ deaths               : int  0 78 187 122 79 6 1 125 5 4 ...
 $ total_pop            : int  6203 160985 173145 126190 60383 83204 25191 151596 36825 30576 ...
 $ male_pop             : int  3135 78928 83544 61154 29705 40952 12504 76162 18214 15366 ...
 $ female_pop           : int  3068 82057 89601 65036 30678 42252 12687 75434 18611 15210 ...
 $ white_pop            : int  5929 143657 108627 115206 57523 71549 23895 129519 34245 29070 ...
 $ black_pop            : int  64 4091 41729 1621 285 2596 214 4425 337 272 ...
 $ asian_pop            : int  32 6144 3459 2436 579 1670 161 6690 711 186 ...
 $ hispanic_pop         : int  83 3542 11820 37

In [32]:
#check

# Convert state names to lowercase and count unique values
unique_state_names <- dataset_plus_census %>%
  mutate(state_name = tolower(state)) %>%  # Convert to lowercase
  summarize(unique_count = n_distinct(state_name))
print(unique_state_names)

# count of state FIPs codes
n_distinct(dataset_plus_census$state_fips_code)

# Convert county names to lowercase, remove "count" from the names, and count unique values
unique_county_names <- dataset_plus_census %>%
    mutate(county_name = tolower(county_name)) %>%
    mutate(count_name = gsub("county", "", county_name, ignore.case = TRUE)) %>%
    summarize(unique_count = n_distinct(county_name))
print(unique_county_names)

  unique_count
1           51


[1] 51

  unique_count
1         1878


There are 51 different states with their corresponding FIPs codes, and 1878 counties in the dataset.

## Chosen Variables

We looked at all of the 31 feature variables that could be correlated to confirmed cases or deaths in project 1. Based on the observations made, following were the variables that seem to be highly correlated with confirmed cases.
- Race
- Income range
- Age group

Therefore, these variables are chosen to peform various types of cluster analysis.

In [37]:
# Basic statistics of the chosen variables
str(dataset_plus_census %>% select(white_pop, black_pop, asian_pop, hispanic_pop, amerindian_pop, other_race_pop,
                                  income_less_50K, income_50K_100K, income_100K_150K, income_150K_more,
                                  male_0_20, male_21_49, male_50_above, female_0_20, female_21_49, female_50_above))

'data.frame':	3142 obs. of  16 variables:
 $ white_pop       : int  5929 143657 108627 115206 57523 71549 23895 129519 34245 29070 ...
 $ black_pop       : int  64 4091 41729 1621 285 2596 214 4425 337 272 ...
 $ asian_pop       : int  32 6144 3459 2436 579 1670 161 6690 711 186 ...
 $ hispanic_pop    : int  83 3542 11820 3769 951 4623 408 7860 795 460 ...
 $ amerindian_pop  : int  20 374 967 1047 130 290 208 38 94 65 ...
 $ other_race_pop  : int  6 240 377 260 65 159 0 336 57 9 ...
 $ income_less_50K : int  1650 24184 27659 16334 9365 12051 4803 16508 5965 6387 ...
 $ income_50K_100K : int  802 20318 21746 14204 8472 10443 3240 16815 5122 3736 ...
 $ income_100K_150K: int  193 11615 8704 9953 3886 6530 1411 10705 2363 1339 ...
 $ income_150K_more: int  75 8789 5272 9117 2856 6397 947 10850 1251 632 ...
 $ male_0_20       : int  677 20796 24873 15755 6834 9057 3293 21806 4747 4099 ...
 $ male_21_49      : int  900 32947 31121 20438 9663 15262 4659 28436 6130 5144 ...
 $ male_50_above  

## K-means (Euclidean distance)

### Clustering Tendency

#### Visual Analysis for Cluster Tendency Assessment (VAT)

#### Hopkins statistic

### Inspect the clusters

### Cluster Evaluation (Unsupervised Cluster Evaluation)

### Elbow Method (Compute within-cluster sum of squares)

### Average Silhouette Width

### Dunn Index

## Agglomerative Hierarchical Clustering (Euclidean distance)

## Third Clustering

## Fourth Clustering